### ISSUE TITLE
Spec Failures

### ISSUE THREAD

**Issue by tibbon** [clearhaus/pedicel#8]
I'm seeing four failures here. Ruby 2.3.7

```
~/C/pedicel ❯❯❯ bundle                                                                                                                                                    master
Fetching gem metadata from https://rubygems.org/........
Resolving dependencies...
Using bundler 1.16.1
Using coderay 1.1.2
Using concurrent-ruby 1.0.5
Using diff-lcs 1.3
Fetching dry-configurable 0.7.0
Installing dry-configurable 0.7.0
Fetching dry-container 0.6.0
Installing dry-container 0.6.0
Fetching dry-core 0.4.5
Installing dry-core 0.4.5
Fetching dry-equalizer 0.2.1
Installing dry-equalizer 0.2.1
Fetching dry-logic 0.4.2
Installing dry-logic 0.4.2
Fetching inflecto 0.0.2
Installing inflecto 0.0.2
Fetching dry-types 0.12.2
Installing dry-types 0.12.2
Fetching dry-validation 0.11.1
Installing dry-validation 0.11.1
Using method_source 0.9.0
Using pedicel 0.0.2 from source at `.`
Using pry 0.11.3
Using rspec-support 3.7.1
Using rspec-core 3.7.1
Using rspec-expectations 3.7.0
Using rspec-mocks 3.7.0
Using rspec 3.7.0
Bundle complete! 3 Gemfile dependencies, 20 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
~/C/pedicel ❯❯❯ rspec                                                                                                                                                     master
FFF.F..

Failures:

  1) basic test works
     Failure/Error: cipher.iv_len = 16

     NoMethodError:
       undefined method `iv_len=' for #<OpenSSL::Cipher:0x00007fe4b0d89bd0>
       Did you mean?  iv_len
     # ./spec/spec_helper.rb:74:in `encrypted_with'
     # ./spec/some_spec.rb:14:in `block (2 levels) in <top (required)>'

  2) basic test works
     Failure/Error: @ca_cert.sign(@ca_key, OpenSSL::Digest::SHA256.new)

     NoMethodError:
       undefined method `private?' for #<OpenSSL::PKey::EC:0x00007fe4b0d83118>
       Did you mean?  private_key?
     # ./spec/spec_helper.rb:148:in `sign'
     # ./spec/spec_helper.rb:148:in `new_ca'
     # ./spec/spec_helper.rb:118:in `initialize'
     # ./spec/some_spec.rb:26:in `new'
     # ./spec/some_spec.rb:26:in `block (2 levels) in <top (required)>'

  3) Spreedly's gala Ruby library's test case decryption with private key and certificate decrypts correctly
     Failure/Error: raise SignatureError, "signature too old; signed #{-diff.to_i}s ago"

     Pedicel::SignatureError:
       signature too old; signed 14400s ago
     # ./lib/pedicel/base.rb:192:in `verify_signed_time'
     # ./lib/pedicel/base.rb:138:in `verify_signature'
     # ./lib/pedicel/ec.rb:28:in `decrypt'
     # ./spec/spreedly_spec.rb:138:in `block (3 levels) in <top (required)>'

  4) Spreedly's gala Ruby library's test case symmetric key can decrypt using the symmetric key
     Failure/Error: raise SignatureError, "signature too old; signed #{-diff.to_i}s ago"

     Pedicel::SignatureError:
       signature too old; signed 14400s ago
     # ./lib/pedicel/base.rb:192:in `verify_signed_time'
     # ./lib/pedicel/base.rb:138:in `verify_signature'
     # ./lib/pedicel/ec.rb:28:in `decrypt'
     # ./spec/spreedly_spec.rb:156:in `block (3 levels) in <top (required)>'

Finished in 0.0178 seconds (files took 0.29029 seconds to load)
7 examples, 4 failures

Failed examples:

rspec ./spec/some_spec.rb:8 # basic test works
rspec ./spec/some_spec.rb:25 # basic test works
rspec ./spec/spreedly_spec.rb:137 # Spreedly's gala Ruby library's test case decryption with private key and certificate decrypts correctly
rspec ./spec/spreedly_spec.rb:155 # Spreedly's gala Ruby library's test case symmetric key can decrypt using the symmetric key
```

**Comment by ct-clearhaus** [comment]
Hi @tibbon, thanks for reporting!

1) The specs will currently only work for Ruby >= 2.4 since `#iv_len=` does not exist in the stdlib OpenSSL. If you use the OpenSSL gem though, it'll work.
Using the OpenSSL requires `gem 'openssl'` before the first `require 'openssl'`. This will happen automatically when using Bundler with version >= 1.14.5. Bundler versions prior to this will `require 'openssl'` before you can have it `gem 'openssl'`; cf. https://github.com/bundler/bundler/blob/master/CHANGELOG.md#1145-2017-02-22 and https://github.com/bundler/bundler/compare/v1.14.4...v1.14.5#diff-3c4d1ce66e392bb28ba4e2e2e86e6e02R1125 and https://github.com/bundler/bundler/pull/5388.

2) This is due to OpenSSL; the API for EC simply differs from the other PKey APIs. The rabbit hole opens up via https://bugs.ruby-lang.org/issues/6567.

3) and 4). I'll have to look into these.

We're working on improving the specs and adjusting the API also. The specs will be improved using https://github.com/clearhaus/pedicel-pay (which is also see quite a diff in a PR). The work is happening in https://github.com/clearhaus/pedicel/pull/7 🙂
As for 2), I have this locally while working on #7:
```ruby
# Monkey-patch to make OpenSSL::X509::Certificate#sign work.
class OpenSSL::PKey::EC
  def private?
    private_key?
  end
end
```

Perhaps https://rubygems.org/gems/pedicel was pushed a bit early 🙂

**Comment by tibbon** [comment]
Under 2.4.4 I get two failures:

```
~/C/pedicel ❯❯❯ rspec                                                                                                                                               ✘ 1 master ◼
..F.F..

Failures:

  1) Spreedly's gala Ruby library's test case decryption with private key and certificate decrypts correctly
     Failure/Error: raise SignatureError, "signature too old; signed #{-diff.to_i}s ago"

     Pedicel::SignatureError:
       signature too old; signed 14400s ago
     # ./lib/pedicel/base.rb:192:in `verify_signed_time'
     # ./lib/pedicel/base.rb:138:in `verify_signature'
     # ./lib/pedicel/ec.rb:28:in `decrypt'
     # ./spec/spreedly_spec.rb:138:in `block (3 levels) in <top (required)>'

  2) Spreedly's gala Ruby library's test case symmetric key can decrypt using the symmetric key
     Failure/Error: raise SignatureError, "signature too old; signed #{-diff.to_i}s ago"

     Pedicel::SignatureError:
       signature too old; signed 14400s ago
     # ./lib/pedicel/base.rb:192:in `verify_signed_time'
     # ./lib/pedicel/base.rb:138:in `verify_signature'
     # ./lib/pedicel/ec.rb:28:in `decrypt'
     # ./spec/spreedly_spec.rb:156:in `block (3 levels) in <top (required)>'

Finished in 0.01486 seconds (files took 0.15424 seconds to load)
7 examples, 2 failures

Failed examples:

rspec ./spec/spreedly_spec.rb:137 # Spreedly's gala Ruby library's test case decryption with private key and certificate decrypts correctly
rspec ./spec/spreedly_spec.rb:155 # Spreedly's gala Ruby library's test case symmetric key can decrypt using the symmetric key
```

**Comment by tibbon** [comment]
Under 2.3.7 I've required openssl at the top of the spec helper, and added that monkey patch just afterward. I'm happy to test things and see where they break on different systems, and just appreciate the work you've done in putting together these libraries and making them available. Getting some slightly different errors now: 

```
~/C/pedicel ❯❯❯ rspec spec                                                                                                                                                                                                                                       ✘ 1 master ✱ ◼
FFF.F..

Failures:

  1) basic test works
     Failure/Error: cipher.iv_len = 16

     NoMethodError:
       undefined method `iv_len=' for #<OpenSSL::Cipher:0x00007fd704deb088>
       Did you mean?  iv_len
     # ./spec/spec_helper.rb:81:in `encrypted_with'
     # ./spec/some_spec.rb:14:in `block (2 levels) in <top (required)>'

  2) basic test works
     Failure/Error: cipher.iv_len = 16

     NoMethodError:
       undefined method `iv_len=' for #<OpenSSL::Cipher:0x00007fd704de1a88>
       Did you mean?  iv_len
     # ./spec/spec_helper.rb:81:in `encrypted_with'
     # ./spec/some_spec.rb:39:in `block (2 levels) in <top (required)>'

  3) Spreedly's gala Ruby library's test case decryption with private key and certificate decrypts correctly
     Failure/Error: raise SignatureError, 'root certificate has not issued intermediate certificate'

     Pedicel::SignatureError:
       root certificate has not issued intermediate certificate
     # ./lib/pedicel/base.rb:161:in `verify_root_certificate'
     # ./lib/pedicel/base.rb:124:in `verify_signature'
     # ./lib/pedicel/ec.rb:28:in `decrypt'
     # ./spec/spreedly_spec.rb:138:in `block (3 levels) in <top (required)>'

  4) Spreedly's gala Ruby library's test case symmetric key can decrypt using the symmetric key
     Failure/Error: raise SignatureError, 'root certificate has not issued intermediate certificate'

     Pedicel::SignatureError:
       root certificate has not issued intermediate certificate
     # ./lib/pedicel/base.rb:161:in `verify_root_certificate'
     # ./lib/pedicel/base.rb:124:in `verify_signature'
     # ./lib/pedicel/ec.rb:28:in `decrypt'
     # ./spec/spreedly_spec.rb:156:in `block (3 levels) in <top (required)>'

Finished in 0.01269 seconds (files took 0.19509 seconds to load)
7 examples, 4 failures

Failed examples:

rspec ./spec/some_spec.rb:8 # basic test works
rspec ./spec/some_spec.rb:25 # basic test works
rspec ./spec/spreedly_spec.rb:137 # Spreedly's gala Ruby library's test case decryption with private key and certificate decrypts correctly
rspec ./spec/spreedly_spec.rb:155 # Spreedly's gala Ruby library's test case symmetric key can decrypt using the symmetric key
```

**Comment by ct-clearhaus** [comment]
Hi @tibbon, I understand the errors you hit when running with Ruby 2.3.7; these will be fixed by #7.

The errors you see with Ruby 2.4.4, I cannot reproduce:
```
$ docker run --rm -it -v $PWD:/pedicel ruby:2.4.4 bash
root@53567132dddc:/# cd pedicel/
root@53567132dddc:/pedicel# bundle
Fetching gem metadata from https://rubygems.org/........
Using bundler 1.16.1
Fetching coderay 1.1.2
Installing coderay 1.1.2
Fetching concurrent-ruby 1.0.5
Installing concurrent-ruby 1.0.5
Fetching diff-lcs 1.3
Installing diff-lcs 1.3
Fetching dry-configurable 0.7.0
Installing dry-configurable 0.7.0
Fetching dry-container 0.6.0
Installing dry-container 0.6.0
Fetching dry-core 0.4.5
Installing dry-core 0.4.5
Fetching dry-equalizer 0.2.1
Installing dry-equalizer 0.2.1
Fetching dry-logic 0.4.2
Installing dry-logic 0.4.2
Fetching inflecto 0.0.2
Installing inflecto 0.0.2
Fetching dry-types 0.12.2
Installing dry-types 0.12.2
Fetching dry-validation 0.11.1
Installing dry-validation 0.11.1
Fetching method_source 0.9.0
Installing method_source 0.9.0
Using pedicel 0.0.2 from source at `.`
Fetching pry 0.11.3
Installing pry 0.11.3
Fetching rspec-support 3.7.1
Installing rspec-support 3.7.1
Fetching rspec-core 3.7.1
Installing rspec-core 3.7.1
Fetching rspec-expectations 3.7.0
Installing rspec-expectations 3.7.0
Fetching rspec-mocks 3.7.0
Installing rspec-mocks 3.7.0
Fetching rspec 3.7.0
Installing rspec 3.7.0
Bundle complete! 3 Gemfile dependencies, 20 gems now installed.
Bundled gems are installed into `/usr/local/bundle`
root@53567132dddc:/pedicel# bundle exec rspec
.......

Finished in 0.02032 seconds (files took 0.13759 seconds to load)
7 examples, 0 failures
```

However! 14400s = 4h. Is your local time by any chance UTC+4h? Can I persuade you to run the specs against #7?
